core/arch/riscv/kernel/thread_rv.S

```diff
@@ -0,0 +1,372 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2022-2023 NXP
+ */
+
+#include <asm.S>
+#include <generated/asm-defines.h>
+#include <keep.h>
+#include <kernel/thread.h>
+#include <kernel/thread_private.h>
+#include <mm/core_mmu.h>
+#include <riscv.h>
+#include <riscv_macros.S>
+
+.macro get_thread_ctx res, tmp0
+	lw	\tmp0, THREAD_CORE_LOCAL_CURR_THREAD(tp)
+	la	\res, threads
+1:
+	beqz	\tmp0, 2f
+	addi	\res, \res, THREAD_CTX_SIZE
+	addi	\tmp0, \tmp0, -1
+	bnez	\tmp0, 1b
+2:
+.endm
+
+.macro save_regs, mode
+	addi	sp, sp, -THREAD_TRAP_REGS_SIZE
+.if \mode == TRAP_MODE_USER
+
+	/* Save user thread pointer and load kernel thread pointer */
+	store_xregs sp, THREAD_TRAP_REG_TP, 4, 4
+	load_xregs sp, (THREAD_TRAP_REGS_SIZE - RISCV_XLEN_BYTES), 4, 4
+
+	store_xregs sp, THREAD_TRAP_REG_GP, 3, 3
+
+	/*
+	 * Set the scratch register to 0 such in case of a recursive
+	 * exception thread_trap_vect() knows that it is emitted from kernel.
+	 */
+	csrrw	gp, CSR_XSCRATCH, zero
+	store_xregs sp, THREAD_TRAP_REG_SP, 3, 3
+.option push
+.option norelax
+	la	gp, __global_pointer$
+.option pop
+.endif
+	store_xregs sp, THREAD_TRAP_REG_T3, 28, 31
```

Comment:
@maroueneboubakri  I agree with Jens, ABI name is more readable and I think most of RISC-V people prefer it than x0, x1...
It is worth a try to unify them as ABI name.

Reply:
Is it enough to add defines to bind registers to ABI names ? `#define RISCV_REG_RA	1` and use it as argument for macros store/load_regs ? `load_xregs a0, THREAD_CTX_REG_RA, RISCV_REG_RA, RISCV_REG_SP` (load RA and SP)

Reply:
Yes! That would be better.

Besides, I recommended that `load_xregs/store_xregs` can be directly unfolded to multiple line of code instead of using one-line macro. It would be helpful for debug since we can intuitively know which registers are loaded/stored now.
If you agree about this, it can be done in other patches.

Reply:
@gagachang the reason for unfolding sounds like more of a tools issue at your end. When saving and restoring loads of registers it's quite nice to be able to avoid cluttering the code with repeated code.

Reply:
@jenswi-linaro Yes you are right, the final code can be checked in other ways such as disassembly of final binary. Shorter code here is helpful for maintenance. Thanks for correction